compiler/rustc_errors/src/json.rs

```diff
@@ -138,6 +138,25 @@ impl JsonEmitter {
     pub fn ui_testing(self, ui_testing: bool) -> Self {
         Self { ui_testing, ..self }
     }
+
+    fn emit(&mut self, val: EmitTyped<'_>) -> io::Result<()> {
+        if self.pretty {
+            serde_json::to_writer_pretty(&mut *self.dst, &val)?
+        } else {
+            serde_json::to_writer(&mut *self.dst, &val)?
+        };
+        self.dst.write_all(b"\n")?;
+        self.dst.flush()
+    }
+}
+
+#[derive(Serialize)]
+#[serde(tag = "type", rename_all = "snake_case")]
```

Comment:
The compiler and standard library generally use the word "kind" for naming this type of field:

https://doc.rust-lang.org/stable/std/?search=kind

https://doc.rust-lang.org/nightly/nightly-rustc/?search=kind

Reply:
I chose `type` specifically because, as a keyword, it's not going to conflict with a real field. But I'm fine with `kind` if that's a stronger convention.

Reply:
I updated https://github.com/rust-lang/compiler-team/issues/673 to discuss why I think `type` is actually the best choice here.

Reply:
There is also discussion in the [MCP zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/233931-t-compiler.2Fmajor-changes/topic/Add.20.60type.60.20field.20to.20distinguish.20json.20dia.E2.80.A6.20compiler-team.23673). Given there is two other places already, let's not discuss the question *here* further.